cpp/src/community/louvain.cuh

```diff
@@ -0,0 +1,639 @@
+    //         could pass that in rather than computing d_deg... which
+    //         would save an atomicAdd (synchronization)
+    //
+    thrust::for_each(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_vertices),
+      [d_inc, d_deg, d_offsets, d_indices, d_weights, d_cluster] __device__(vertex_t v) {
+        vertex_t community = d_cluster[v];
+        weight_t increase{0.0};
+        weight_t degree{0.0};
+
+        for (edge_t loc = d_offsets[v]; loc < d_offsets[v + 1]; ++loc) {
+          vertex_t neighbor = d_indices[loc];
+          degree += d_weights[loc];
+          if (d_cluster[neighbor] == community) { increase += d_weights[loc]; }
+        }
+
+        if (degree > weight_t{0.0}) atomicAdd(d_deg + community, degree);
+        if (increase > weight_t{0.0}) atomicAdd(d_inc + community, increase);
+      });
+
+    weight_t Q = thrust::transform_reduce(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_vertices),
+      [d_deg, d_inc, total_edge_weight, resolution] __device__(vertex_t community) {
+        return ((d_inc[community] / total_edge_weight) - resolution *
+                                                           (d_deg[community] * d_deg[community]) /
+                                                           (total_edge_weight * total_edge_weight));
+      },
+      weight_t{0.0},
+      thrust::plus<weight_t>());
+
+    return Q;
+  }
+
+  virtual std::pair<int, weight_t> compute(vertex_t *d_cluster_vec,
+                                           int max_level,
+                                           weight_t resolution)
+  {
+    int num_level{0};
+
+    weight_t total_edge_weight =
+      thrust::reduce(rmm::exec_policy(stream_)->on(stream_), weights_v_.begin(), weights_v_.end());
+
+    weight_t best_modularity = weight_t{-1};
+
+    //
+    //  Initialize every cluster to reference each vertex to itself
+    //
+    thrust::sequence(rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end());
+    thrust::copy(
+      rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end(), d_cluster_vec);
+
+    //
+    //  Our copy of the graph.  Each iteration of the outer loop will
+    //  shrink this copy of the graph.
+    //
+    GraphCSRView<vertex_t, edge_t, weight_t> current_graph(offsets_v_.data().get(),
+                                                           indices_v_.data().get(),
+                                                           weights_v_.data().get(),
+                                                           number_of_vertices_,
+                                                           number_of_edges_);
+
+    current_graph.get_source_indices(src_indices_v_.data().get());
+
+    while (num_level < max_level) {
```

Comment:
See comment on using a `for_each()`, `transform()`, (or, `transform_if()`, `std::any_of()`, `std::all_of()`, `std::none_of()`) , due to the possibility of early break)` on the sequence `{0,...,max_level-1}` rather than a while loop.